drivers/crypto/crypto_mtls_shim.c

```diff
@@ -103,36 +104,55 @@ int mtls_ecb_decrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt)
 
 int mtls_cbc_encrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt, u8_t *iv)
 {
-	int ret;
+	int ret, iv_bytes;
+	u8_t *p_iv, iv_loc[16];
 	mbedtls_aes_context *cbc_ctx = MTLS_GET_CTX(ctx, aes);
 
-	/* Prefix IV to ciphertext */
-	memcpy(pkt->out_buf, iv, 16);
+	if ((ctx->flags & CAP_NO_IV_PREFIX) == 0U) {
+		iv_bytes = 16;
+		memcpy(pkt->out_buf, iv, 16);
+		p_iv = iv;
+	} else {
+		iv_bytes = 0;
+		memcpy(iv_loc, iv, 16);
+		p_iv = iv_loc;
+	}
+
 	ret = mbedtls_aes_crypt_cbc(cbc_ctx, MBEDTLS_AES_ENCRYPT, pkt->in_len,
-				    iv, pkt->in_buf, pkt->out_buf + 16);
+				    p_iv, pkt->in_buf, pkt->out_buf + iv_bytes);
 	if (ret) {
 		LOG_ERR("Could not encrypt (%d)", ret);
 		return -EINVAL;
 	}
 
-	pkt->out_len = pkt->in_len + 16;
+	pkt->out_len = pkt->in_len + iv_bytes;
 
 	return 0;
 }
 
 int mtls_cbc_decrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt, u8_t *iv)
 {
-	int ret;
+	int ret, iv_bytes;
+	u8_t *p_iv, iv_loc[16];
 	mbedtls_aes_context *cbc_ctx = MTLS_GET_CTX(ctx, aes);
 
+	if ((ctx->flags & CAP_NO_IV_PREFIX) == 0U) {
+		iv_bytes = 16;
+		p_iv = iv;
+	} else {
+		iv_bytes = 0;
+		memcpy(iv_loc, iv, 16);
+		p_iv = iv_loc;
+	}
+
 	ret = mbedtls_aes_crypt_cbc(cbc_ctx, MBEDTLS_AES_DECRYPT, pkt->in_len,
-				    iv, pkt->in_buf + 16, pkt->out_buf);
```

Comment:
This shows how weird is the old behavior: it allows to pass in arbitrary `iv` pointer, but still assumes there's an IV to skip in the input packet ;-/.

Reply:
I believe that this is embedtls shim's idiosyncrasy. Zephyr's crypto API didn't make any assumptions there.